Brokerages/Bitfinex/BitfinexWebsocketsBrokerage.Messaging.cs

```diff
@@ -0,0 +1,214 @@
+                    }
+                    else if (term == "ws")
+                    {
+                        //wallet
+                        var data = raw[2].ToObject(typeof(string[][]));
+                        PopulateWallet(data);
+                    }
+                    else if (_channelId.ContainsKey(id) && _channelId[id] == "ticker")
+                    {
+                        //ticker
+                        PopulateTicker(e.Data);
+                        return;
+                    }
+                }
+                else if (raw.channel == "ticker" && raw.@event == "subscribed")
+                {
+                    if (!this._channelId.ContainsKey((int)raw.chanId))
+                    {
+                        this._channelId[(int)raw.chanId] = "ticker";
+                    }
+                }
+                else if (raw.chanId == 0)
+                {
+                    if (raw.status == "FAIL")
+                    {
+                        throw new Exception("Failed to authenticate with ws gateway");
+                    }
+                    Log.Trace("Successful wss auth");
+                }
+                else if (raw.@event == "info" && raw.code == "20051")
+                {
+                    //hard reset
+                    this.Reconnect();
+                }
+                else if (raw.@event == "info" && raw.code == "20061")
+                {
+                    //soft reset
+                    UnAuthenticate();
+                    Unsubscribe(null, null);
+                    Subscribe(null, null);
+                    Authenticate();
+                }
+
+                Log.Trace(e.Data);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, string.Format("Parsing wss message failed. Data: {0}", e.Data));
+            }
+        }
+        
+        private void PopulateTicker(string response)
+        {
+            var data = JsonConvert.DeserializeObject<string[]>(response);
+            var msg = new TickerMessage(data);
+            lock (ticks)
+            {
+                ticks.Add(new Tick
+                {
+                    AskPrice = msg.ASK / divisor,
+                    BidPrice = msg.BID / divisor,
+                    AskSize = (long)Math.Round(msg.ASK_SIZE * divisor, 0),
+                    BidSize = (long)Math.Round(msg.BID_SIZE * divisor, 0),
+                    Time = DateTime.UtcNow,
+                    Value = msg.LAST_PRICE / divisor,
+                    TickType = TickType.Quote,
+                    Symbol = symbol,
+                    DataType = MarketDataType.Tick,
+                    Quantity = (int)(Math.Round(msg.VOLUME, 2) * divisor)
+                });
+            }
+        }
+
+        //todo: Currently populated but not used
+        private void PopulateWallet(string[][] data)
+        {
+            if (data.Length > 0)
+            {
+                lock (_cash)
```

Comment:
It's typically better to have a dedicated object for locking, so maybe we could have a `_cashLock` object instead of locking directly on the list